### PR TITLE
Spiders/carp will now pull/move water/welding fuel tanks/canisters slower and won't be able to attack stationary atmospherics equipment

### PR DESCRIPTION
**Pull request by Mothblocks**
## About The Pull Request
Adds 2 new elements, one for slowing down pulling of dangerous objects (dispenser tanks and canisters), and one for preventing hostile attacking of elements in a typecache.

Also updates the obj_flags bitfield 'cause I thought I was gonna use that, but I didn't.

## Why It's Good For The Game
Spiders and space carp are near infinitely respawning player controlled characters with zero reason to care about their life. While this can be seen as a problem in their design, this is something we hit often enough to warrant limiting their capabilities up front.

Pulling and moving of dispensers/canisters is being used instead of outright blocking it. This is because it stresses the importance of environmental combat, meaning spiders in maintenance can still use nearby water/fuel tanks to their advantage, but are heavily disadvantaged from bringing that out to the rest of the station (limiting environmental combat elsewhere).

Preventing them from attacking stationary atmospherics equipment helps prevent carp/spiders from completely sabotaging the SM, inevitably leading to the round ending even if the dragon was taken care of, for instance.

## Changelog

:cl:
balance: Spiders and space carp (from dragons) will now pull and move dispensers, such as water and fuel tanks, as well as canisters slower.
balance: Spiders and space carp (from dragons) can no longer attack stationary atmospherics equipment, such as pipes or air alarms, as well as APCs.
/:cl:

### CODE REVIEW

File: code/datums/elements/nerfed_pulling.dm [hunk 1]
```diff
@@ -0,0 +1,52 @@
+/// This living will be slower when pulling/moving some dangerous objects in its obj_flags
+/datum/element/nerfed_pulling
+	element_flags = ELEMENT_DETACH
+
+	var/static/list/nerf_pulling_of_typecache = typecacheof(list(
```
Comment (Mothblocks) [inline code/datums/elements/nerfed_pulling.dm:5]: Gonna turn this into an argument for a bespoke element, rather than an arbitrary static.

**Comment by Mothblocks** [comment]
look at that id nice

**Comment by Pickle-Coding** [comment]
I think the last part of the changelog got cut or something.

**Comment by Mothblocks** [comment]
i died  to a spider while writing it

**Review by Ghommie — CHANGES_REQUESTED** [review]
Right now, this change will also affect carps not spawned by carp rift such as Cayenne, Lia, the wizard shapeshift chaos magicarp, mega space carps and in general any other player controlled carp mob, be it through sentience potion, event or admin shenanigeans, which actually aren't "_near infinitely respawning player controlled characters with zero reason to care about their life_" for most part and as such should be exempt from it. The same could be said for any spider not spawned from eggs, spiderlings and the spider mutation extract, but they're fragile by design so I don't mind either ways.

I hope you'll resolve this issue.

TL;DR: Please make this change only affect rift carps versus all carps.

**Comment by IndieanaJones** [comment]
Space carp from rifts have a respawn cooldown that stops people from respawning from the same rift unless there's a backup of respawns available (2 or more).  They also had their numbers nerfed.  I guess this makes sense for when SD wins and they are actually infinite, but I'm surprised they're still suicide bombing people beforehand.

Spiders, on the other hand, really do need this.

**Comment by Mothblocks** [comment]
@Ghommie sure

**Comment by MrMelbert** [comment]
Include APCs, too? they can still sabotage the SM by breaking all the APCs around it

**Comment by Mothblocks** [comment]
I might specifically filter those APCs, but I think *in general* them destroying APCs is *ooookaaaay*. I'll think on it.

**Comment by Mothblocks** [comment]
Well, APCs are still dumb as fuck to fix (since I haven't made my changes to them yet), so I guess it's fair enough.

**Comment by Mothblocks** [comment]
i think its a tad weird that carp that can push/pull stuff easily and carp that cant will be undiscernable but i dont mind that much

**Comment by LemonInTheDark** [comment]
APC DESTROYED, MISSION ACCOMPLISHED

File: code/datums/elements/nerfed_pulling.dm [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+/datum/element/nerfed_pulling/proc/on_updating_pull_movespeed(mob/living/source)
+	SIGNAL_HANDLER
+
+	if (!will_slow_down(source.pulling))
+		source.remove_movespeed_modifier(/datum/movespeed_modifier/nerfed_pull)
+		return
+
+	source.add_movespeed_modifier(/datum/movespeed_modifier/nerfed_pull)
+
+/datum/element/nerfed_pulling/proc/will_slow_down(datum/input)
```
Comment (Timberpoes) [inline code/datums/elements/nerfed_pulling.dm:46]: This proc can be called from `on_updating_pull_movespeed()`. This proc has two code execution paths, one for updating when the mob is pulling something and one for when the mob isn't pulling anything.

You added `SEND_SIGNAL(src, COMSIG_LIVING_UPDATING_PULL_MOVESPEED)` at the beginning of this proc when `pulling` can be null (for example, when `/mob/living/stop_pulling()` does its parent call to `/atom/movable/proc/stop_pulling()` which can null `pulling`, then calls `on_updating_pull_movespeed()` after).

This can lead to `source.pulling` being null and you passing `null` as the value for `input`. It will then runtime on `return typecache[input.type]`.

**Review by Timberpoes — CHANGES_REQUESTED** [review]
Just this one thing I've identified through code logic.

File: code/datums/elements/nerfed_pulling.dm (hunk 2 above)
Comment (Mothblocks) [reply code/datums/elements/nerfed_pulling.dm:46]: :+1: Good catch, the old proc for this checked `isobj`, which would've failed that.

File: code/datums/elements/nerfed_pulling.dm [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+	return typecache[input.type]
```
Comment (Mothblocks) [inline code/datums/elements/nerfed_pulling.dm:47]:
```suggestion
	return !isnull(input) && typecache[input.type]
```

**Review by Timberpoes — APPROVED** [review]
